### ISSUE TITLE
Are you plan using OpenSC on iOS device?

### ISSUE THREAD

**Issue by FeitianSmartcardReader** [OpenSC/OpenSC#626]
Hi Guys

There more and more people using mobile device, so are you plan using OpenSC on mobile? android and iOS?

**Comment by frankmorgner** [comment]
Android works, iOS should work. A problem that always needs to be solved independant from OpenSC is the actual communication with the reader.

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner Thanks for your reply. We do publish our bluetooth reader driver source code on github which based on CCID protocol -> https://github.com/FeitianSmartcardReader/bR301_iOS/tree/master/libs/source/src, if there have a chance, we can do some contribution let opensc support iOS.

**Comment by frankmorgner** [comment]
You should wrap your reader into a standalone version of libpcsclite that is accessed by OpenSC through the standard PC/SC interface. For example, I created one that communicates through a socket to the reader, for example, see https://frankmorgner.github.io/vsmartcard/virtualsmartcard/README.html. If you want to reuse my library, you simply connect to the socket from your iOS-based reader implementation.

**Comment by frankmorgner** [comment]
## TL;DR
Try combining [Yubico's sample application](https://github.com/Yubico/yubioath-ios) with [OpenSC's CTK extension](https://github.com/frankmorgner/OpenSCToken).

## How can OpenSC access the smart card or smart card reader?
1. With the announcement for iOS16 it reads that it uses the same exact API as on macOS 12 and later. OpenSC for CTK (on macOS) is available [here](https://github.com/frankmorgner/OpenSCToken/). I have not seen code examples for a CryptoTokenKit provider on iOS, but it looks like you just need to compile OpenSCToken for iOS and ship the app extension with your app. Internally, OpenSC will use reader-cryptotokenkit.m for mapping the smart card commands to the smart card reader. Yubico has a [nice tutorial and sample code](https://www.yubico.com/blog/how-to-implement-a-cryptotokenkit-extension-on-ios/) on their website.
2. The standard interface on desktops for OpenSC is PC/SC (reader-pcsc.c). So an other option would be to implement your own (statically linked) version of libpcsclite, which implements [this interface](https://pcsclite.apdu.fr/api/group__API.html). In the implementation of your libpcsclite you should then call your proprietary API for accessing the smart card; in case of the Feitian smart card reader that would mean to call the bluetooth API [as described above](https://github.com/FeitianSmartcardReader/bR301_iOS/tree/master/libs/source/src)

## How to access OpenSC from an actual iOS implementation?
0. From the superficial documentation of iOS16 it seems that in case of proper CTK integration, all you have to do is use the CTK API to access your smart card, just as you would have done on macOS.
1. you could use (a statically linked version of OpenSC's PKCS#11 interface. The interface is described [here](https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.html) and there are numerous examples on how to access this API. OpenSC provides `pkcs11-tool`, for example.
2. you could directly access the card by using some more low level operations defined in `opensc.h`, which only depend on libopensc. Examples for this API are given in `pkcs15-tool`.

You can link OpenSC statically to all dependencies. All dependencies are C-Libraries, and can thus be compiled with XCode for iOS.

**Comment by xaqfan** [comment]
Sorry, I mean that how do I compile it?
Use xcode or in terminal with command line?
I tried to compile it in terminal,but failed.

I will use the librarys of opensc on ios/ipados 16.x.


> ## TL;DR
> Try combining [Yubico's sample application](https://github.com/Yubico/yubioath-ios) with [OpenSC's CTK extension](https://github.com/frankmorgner/OpenSCToken).
> 
> ## How can OpenSC access the smart card or smart card reader?
> 1. With the announcement for iOS16 it reads that it uses the same exact API as on macOS 12 and later. OpenSC for CTK (on macOS) is available [here](https://github.com/frankmorgner/OpenSCToken/). I have not seen code examples for a CryptoTokenKit provider on iOS, but it looks like you just need to compile OpenSCToken for iOS and ship the app extension with your app. Internally, OpenSC will use reader-cryptotokenkit.m for mapping the smart card commands to the smart card reader. Yubico has a [nice tutorial and sample code](https://www.yubico.com/blog/how-to-implement-a-cryptotokenkit-extension-on-ios/) on their website.
> 2. The standard interface on desktops for OpenSC is PC/SC (reader-pcsc.c). So an other option would be to implement your own (statically linked) version of libpcsclite, which implements [this interface](https://pcsclite.apdu.fr/api/group__API.html). In the implementation of your libpcsclite you should then call your proprietary API for accessing the smart card; in case of the Feitian smart card reader that would mean to call the bluetooth API [as described above](https://github.com/FeitianSmartcardReader/bR301_iOS/tree/master/libs/source/src)
> 
> ## How to access OpenSC from an actual iOS implementation?
> 0. From the superficial documentation of iOS16 it seems that in case of proper CTK integration, all you have to do is use the CTK API to access your smart card, just as you would have done on macOS.
> 1. you could use (a statically linked version of OpenSC's PKCS#11 interface. The interface is described [here](https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.html) and there are numerous examples on how to access this API. OpenSC provides `pkcs11-tool`, for example.
> 2. you could directly access the card by using some more low level operations defined in `opensc.h`, which only depend on libopensc. Examples for this API are given in `pkcs15-tool`.
> 
> You can link OpenSC statically to all dependencies. All dependencies are C-Libraries, and can thus be compiled with XCode for iOS.
